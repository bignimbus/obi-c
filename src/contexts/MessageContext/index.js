import React, { useState, createContext } from 'react';
import PropTypes from 'prop-types';

const MessageContext = createContext();

export const MessageContextProvider = ({ children }) => {
  const [messages, setMessages] = useState([]);

  const enqueueMessage = (message) => {
    setMessages([...messages, message]);
  };

  const dequeueMessage = (message) => {
    setMessages(
      messages
        .filter(m => m !== message),
    );
  };

  return (
    <MessageContext.Provider
      value={{
        messages,
        enqueueMessage,
        dequeueMessage,
      }}
    >
      { children }
    </MessageContext.Provider>
  );
};

MessageContextProvider.propTypes = {
  children: PropTypes.node.isRequired,
};

export default MessageContext;
