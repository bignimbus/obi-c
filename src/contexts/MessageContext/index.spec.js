import React, { useContext } from 'react';
import renderer, { act } from 'react-test-renderer';
import MessageContext, { MessageContextProvider } from '.';

const renderComponentWithContext = Child => renderer.create(
  <MessageContextProvider>
    <Child />
  </MessageContextProvider>,
);

describe('MessageContextProvider', () => {
  it('should provide a context', () => {
    const TestChildComponent = ({ testValue }) => <div>{ JSON.stringify(testValue.messages) }</div>;
    const TestComponent = () => {
      const contextValue = useContext(MessageContext);
      return <TestChildComponent testValue={contextValue} />;
    };
    const element = renderComponentWithContext(TestComponent);
    const instance = element.root.findByType(TestChildComponent);
    const { messages, enqueueMessage, dequeueMessage } = instance.props.testValue;
    expect(messages).toEqual([]);
    expect(element.toJSON()).toMatchSnapshot();
    const stubMessage = { foo: 'bar' };
    act(() => {
      enqueueMessage(stubMessage);
    });
    expect(element.toJSON()).toMatchSnapshot();
    act(() => {
      dequeueMessage(stubMessage);
    });
    expect(element.toJSON()).toMatchSnapshot();
  });
});
