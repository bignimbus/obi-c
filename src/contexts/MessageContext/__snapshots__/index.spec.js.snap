// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MessageContextProvider should provide a context 1`] = `
<div>
  []
</div>
`;

exports[`MessageContextProvider should provide a context 2`] = `
<div>
  [{"foo":"bar"}]
</div>
`;

exports[`MessageContextProvider should provide a context 3`] = `
<div>
  []
</div>
`;
